PEP 669: Low Impact Instrumentation and Monitoring for CPython. (#2070)
python/peps: pep-0669.rst

PEP: 669
Title: Low Impact Instrumentation and Monitoring for CPython
Author: Mark Shannon <mark@hotpy.org>
Status: Draft
Type: Standards Track
Content-Type: text/x-rst
Created: 18-Aug-2021
Post-History: 13-Sep-2021

Abstract
========

Using a profiler or debugger in CPython can have a severe impact on
performance. Slowdowns by an order of magnitude are not uncommon.
It does not have this bad.

This PEP proposes an API for instrumentation and monitoring of Python
programs running on CPython that will enable the insertion of instrumentation
and monitoring at low cost.

Using the new API, code run under a debugger on 3.11 should easily outperform
code run without a debugger on 3.10.

Profiling will still slow down execution, but by much less than in 3.10.

Motivation
==========

Developers should not have to pay an unreasonable cost to use debuggers,
profilers and other similar tools.

C++ and Java developers expect to be able to run a program at full speed
(or very close to it) under a debugger.

> Python developers should expect that too.

Author: It is the "motivation" section; it  should be motivating  :slightly_smiling_face:
I've toned down the abstract to compensate.

> Python developers should expect that too.

Reviewer: For those who think that `pdb` is the peak of debugging, and don't suffer from any performance issues with that, is there any motivation to support the change?

I'm obviously in agreement that a better approach would be great, just concerned that there may be push back from people who don't see the value when a few lines here about the current state of things (maybe the frequent checks for tracing?) would probably motivate things.

Rationale
=========

The quickening mechanism provided by PEP 659 provides a way to dynamically
modify executing Python bytecode. These modifications have no cost beyond
the parts of the code that are modified and a relatively low cost to those
parts that are modified. We can leverage this to provide an efficient
mechanism for instrumentation and monitoring that was not possible in 3.10
or earlier.

Specification
=============

There are two parts to this specification, instrumentation and monitoring.

Instrumentation occurs early in a program's life cycle and persists through
the lifetime of the program. It is expected to be pervasive, but fixed.
Instrumentation is designed to support profiling and coverage tools that 

> expect to be active for the entire lifetime of the program.

Author: To profile calls to a specific function `foo`:
```Python
instrumentation.instrument(foo.__code__, instrumentation.ENTER)
instrumentation.register(instrumentation.ENTER, profiler_func)
```

Monitoring can occur at any point in the program's life and be applied
anywhere in the program. Monitoring points are expected to few.

> The capabilities of monitoring are a superset of that of profiling,

Author: Debugging is a form of monitoring. The debugger is monitoring the program being run.
Any ideas for a better term than monitoring"? I don't want to use "debugging" as that is just a particular use case.

> The capabilities of monitoring are a superset of that of profiling,

Reviewer 2: I think the terms are fine, all I'm asking for is that you clarify the mapping between the two concepts and the three or more use cases. 'Monitoring is a superset of profiling" doesn't do that for me, but "A profiler would use the monitoring API" would. Except your example of profiling a function (in the comment above) uses instrumentation, not profiling. So I'm still confused about the relationship between instrumentation and monitoring, and between those and the use cases.

but bulk insertion of monitoring points will be *much* more
expensive than insertion of instrumentation.

Both instrumentation and monitoring is performed by insertion of
checkpoints in a code object.

Checkpoints
-----------

A checkpoint is simply a point in code defined by a 

> ``(codeobject, offset)`` pair.

Reviewer 2: Are offsets measured in bytes or instructions?

Every time a checkpoint is reached, the registered callable is called.

Instrumentation
---------------

Instrumentation supports the bulk insertion of checkpoints, but does not
allow insertion or removal of checkpoints after code has started to execute.

The events are::

  * BRANCH: A conditional branch is reached. 
  * JUMPBACK: A backwards, unconditional branch is reached.
  * ENTER: A Python function is entered.
  * EXIT: A Python function exits normally (without an exception).
  * UNWIND: A Python function exits with an unhandled exception.
  * C_CALL: A call to any object that is not a Python function.
  * C_RETURN: A return from any object that is not a Python function.
  * RAISE: An exception is raised.

>   * EXCEPT: An exception is handled.

Author: > Does the checking whether the raised exception matches RuntimeError count as "handling"?

Yes. These are VM level events. I've changed the wording to :
```EXCEPT: Control is transferred to an exception handler.```

For each ``ENTER`` event there will be a corresponding
``EXIT`` or ``UNWIND`` event.
For each ``C_CALL`` event there will be a corresponding
``C_RETURN`` or ``RAISE`` event.

All events are integer powers of two and can be bitwise or-ed together to
instrument multiple events.

Instrumenting code objects
''''''''''''''''''''''''''

Code objects can be instrumented by calling::

>   instrumentation.instrument(codeobject, events)

Author: > For code objects containing other code objects (e.g. nested functions/classes, lambdas, comprehensions), does this affect the sub-objects?

No. Just the code object specified.

>   instrumentation.instrument(codeobject, events)

Reviewer 2: That's worth calling out in the text as a clarification then.

>   instrumentation.instrument(codeobject, events)

Reviewer: Is this a new module? Why not put it in `sys`? (Even as `sys.intrumentation...`)

Code objects must be instrumented before they are executed.

> An exception will be raised if the code object has been executed before it

Reviewer: It said earlier that quickening lets us modify "executing" bytecode. Is that not actually the case?

And if so, when do we have a chance to instrument new code objects? Do they all have to be done on creation? Or can we do it on first call?

> is instrumented.

Reviewer 3: I think I had misread this PEP... I thought that only by using `instrumentation.register(instrumentation.ENTER, profiler_func)` I'd start to listen to any function call and `instrumentation.register(instrumentation.UNWIND, profiler_func)` would make it listen to any unhandled exception, but now, rereading, I'm under the impression that I'd just listen to function calls for code objects previously instrumented and it wouldn't even be possible to change it for code objects already running... 

I think that the PEP could make that clearer.

Now, if it's really the case that:

- there's no callback to listen to any call just prior to a code object/frame being executed 
- there's no way to listen to any events for code objects/frames already running

I'd say I'm -1 on the PEP as it'd just not be usable for many use cases (the happy path where this API would work would be too narrow to be usable)... 

For instance, it's **very** common for users to attach to a running program and this PEP wouldn't support that at all and it'd be next to impossible to get a hold of all the code objects that'd need to be instrumented before they're actually run even on the case where the program is started with the debugger in place (an import hook to instrument code objects isn't enough as there are many corner cases where it wouldn't be able to get it).

Now, if I misunderstood it, then I think the PEP should make it clearer that those are in fact supported and which API would be used for that...

Register callback functions for instrumentation
'''''''''''''''''''''''''''''''''''''''''''''''

To register a callable for events call::

  instrumentation.register(event, func)

Functions can be unregistered by calling
``instrumentation.register(event, None)``.

> Callback functions can be registered at any time.

Author: Events are always active unless explicitly disabled. Callbacks can be any callable, including Python code.
Instrumenting your instrumentation callback function will result in a recursion error. So you can't profile a profiler.

You can add monitor checkpoints to the callback function for instrumentation. It will be possible to debug a profiler.

> Callback functions can be registered at any time.

Reviewer 2: Ah, this makes sense now I understand that each code object needs to be instrumented separately. Nevertheless your clarifications in the comments would be helpful if added to the text.

> Callback functions can be registered at any time.

Reviewer: Seconding the question about whether they're global/per-interpreter/per-thread/per-...whatever.

Right now, people hit [issues](https://twitter.com/willmcgugan/status/1435691567306715143?s=20), [confusion](https://bugs.python.org/issue23163) and sometimes [ideas](https://bugs.python.org/issue41571) due to settrace's single-threadedness (particularly when you start tracing an already-running application and don't have a chance to force your trace into every thread, which is something we also saw regularly building the debugger for VS). The main problem being that tracing on one thread doesn't give you any way to trace - or even pause execution of - the others.

e.g. once you hit a breakpoint in one thread, the _first_ thing you do that releases the GIL is going to let other threads start executing. Having some way to also trigger an event in the context of every thread would let a thread-aware debugger control which of them can keep executing. (Or maybe some other kind of approach makes more sense, that's just an idea that jumps out at me.) It doesn't seem to be essential to this PEP for this to be supported, but it's a good opportunity to add it while we're defining new events.

At the very least, having clear statements about the threading behaviour for what this PEP _does_ add are essential.

Callback function arguments
'''''''''''''''''''''''''''

> When an event occurs the registered function will be called.

Reviewer: And events only occur if they were instrumented?

The arguments provided are as follows:

> * BRANCH: ``func(code: CodeType, offset: int, taken:bool)``

Reviewer: Is the code object an important parameter? The frame is generally more useful (and implicitly includes the code object).

Reviewer: Then again, I guess for a profiler the code object is enough as it's likely maintaining its own call stacks. Debuggers will need to walk the frame even if callers weren't instrumented, and so having them do an extra call to start walking the stack is better than preemptively passing it in.

If that's the case, a sentence explaining it would be nice, so the next reader doesn't have to guess.

Reviewer 3: +1 to the case where the frame is needed for debuggers (i.e.: this PEP should have some note stating the recommended way to get the frame related to the instrumentation notification).

-- as a note, if performance-wise it'd be the same, I'd say that receiving the frame which contains the code object would be better. If there's some performance penalty, then another way to get it would also be reasonable since it's not always needed (but the recommended way to get it should be explicit in the PEP).

* JUMPBACK: ``func(code: CodeType, offset: int)``
* ENTER: ``func(code: CodeType, offset: int)``
* EXIT: ``func(code: CodeType, offset: int)``
* C_CALL: ``func(code: CodeType, offset: int, value: object)``
* C_RETURN: ``func(code: CodeType, offset: int, value: object)``

> * C_EXCEPT: ``func(code: CodeType, offset: int, exception: BaseException)``

Reviewer 3: I believe that both the exception as well as the traceback would be important in the `C_EXCEPT`, `RAISE`, `EXCEPT` as well as the `UNWIND` cases.

As a note, if the idea is using `exception.__traceback__`, it'd be nice if if that's noted in the PEP... as a note, I'm not sure if `exception.__traceback__` is always available -- is it possible that some exception raised -- say some custom exception object from C/C++ doesn't have it?  (in which case it'd need to be passed as a parameter?)

* RAISE: ``func(code: CodeType, offset: int, exception: BaseException)``
* EXCEPT: ``func(code: CodeType, offset: int)``
* UNWIND: ``func(code: CodeType)``

Monitoring
----------

Monitoring allows checkpoints to be inserted or removed at any
point in the program's execution.

The following functions are provided to insert monitoring points::

  instrumentation.insert_monitors(codeobject, *offsets)
  instrumentation.remove_monitors(codeobject, *offsets)
  instrumentation.monitor_off(codeobject, offset)

>   instrumentation.monitor_on(codeobject, offset)

Reviewer 2: Is `monitor_off(co, offs)` equivalent to `insert_monitors(co, [offs])`? If so, why bother with the single-offset API at all? Certainly a user of this API can be trusted to listify the argument?

Author: > Is monitor_off(co, offs) equivalent to insert_monitors(co, [offs])?

No. Turning a monitor on or off is a very cheap (<1us) operation. Inserting or removing a monitor is a very expensive operation maybe taking 100s of milliseconds as it may cause cascading de-optimizations.

So, we want to allow multiple monitoring checkpoints to be inserted at once because:

1. Each insertion (or removal) or monitors can be very expensive.
2. Debuggers need to be able to insert multiple check points for lines that have several instructions (e.g. code in finally blocks, and duplicated tests in while loops).

If a user can't listify the arguments, I don't have much hope of them implementing a debugger :slightly_smiling_face:

Reviewer 2: Oh. This important distinction might be easier to follow if you allowed yourself some mention of the (roughly) intended implementation. In my imagination, inserting monitors is roughly equivalent to

1. Specialize the code object if not already specialized
2. For each offset, replace the specialized opcode with a MONITOR opcode
3. Maybe clear the inline cache for that code object?

But then monitor_off() would seem to be equivalent to replacing the MONITOR opcode with the original opcode (or perhaps the specialized variant, if we care).

hat kind of cascading de-optimizations am I missing? Anything that's currently implemented, or is this referring to e.g. generating machine code?

All functions return ``True`` if a monitor checkpoint was present,

> or ``False`` if a monitor checkpoint was not present.

Reviewer 2: This is a little vague. If I insert checkpoints at offsets 10 and 20, and one was present at 10 but not at 20, should it return True or False? Or should it return a list[bool]?

Author: Yes that is ambiguous.
I've changed that part of the API.

Turning on, or off, a non-existent checkpoint is a no-op;

> no exception is raised.

Reviewer 2: But presumably it would return False, right?

Author: Yes

To register a callable for monitoring function events call::

  instrumentation.monitor_register(func)

The callback function will be called with the code object and offset as arguments::

  func(code: CodeType, offset: int)

For optimizing virtual machines, such as future versions of CPython
(and ``PyPy`` should they choose to support this API), a call to
``insert_monitors`` and ``remove_monitors`` in a long running program
could be quite expensive, possibly taking 100s of milliseconds as it

> triggers de-optimizations. Repeated calls to ``insert_monitors``

Reviewer 2: Where do you foresee these 100s of ms being spent? Tracking down all the affected code objects? Rewriting machine code? Clearing the inline cache data? What else?

Author: Yes those sort of things. We might need to scan *all* compiled code to see what needs to be de-optimized, then do quite a lot of clean up.

I would expect this to only happen in an interactive debugger, where a 500ms pause is barely noticeable.

Reviewer 2: (I realize I asked the same question again before reading your reply. But I'm still not sure of the scope of the deoptimization. All your APIs are specific to a code object, and I can't imagine deoptimizing a single code object taking more than a few msec. So there must be some kind of optimization you are planning that spans multiple code objects. What? Is there an issue in the faster-cpython/ideas tracker about this?)

and ``remove_monitors``, as may be required in an interactive debugger,
should be relatively inexpensive.

Combining Checkpoints
---------------------

Only one instrumentation checkpoint and one monitoring checkpoint is allowed
per bytecode instruction. It is possible to have both a monitoring and
instrumentation checkpoint on the same instruction; they are independent.

> Monitors will be called before instrumentation if both are present.

Reviewer 2: This confused me, since the instrumentation API has no explicit mention of checkpoints (there are no offsets in the instrumentation API, only code objects). I take it that instrumentation works by replacing specific bytecodes (e.g. certain JUMP instructions for BRANCH and JUMPBACK), and monitoring *also* works by doing that? And the instruction has two flag bits indicating whether it is an instrumentation or monitoring checkpoint or both?

It would seem that the APIs already imply that only one instrumentation checkpoint can exist per instruction, and only one monitoring checkpoint. So this paragraph only adds in which order they will be called.

Backwards Compatibility
=======================

This PEP is fully backwards compatible.

We may seek to remove ``sys.settrace`` in the future once the APIs provided
by this PEP have been widely adopted, but that is for another PEP.

Security Implications
=====================

Allowing modification of running code has some security implications,
but no more than the ability to generate and call new code.

All the functions listed above will trigger audit hooks.

Implementation
==============

> The implementation of this PEP will be built on top of PEP 659 quickening.

Reviewer 2: ```suggestion
The implementation of this PEP will be built on top of PEP 659 (Specializing Adaptive Interpreter).
```

Instrumentation or monitoring of a code object will cause it to be quickened.
Checkpoints will then be implemented by inserting one of several special
``CHECKPOINT`` instructions into the quickened code. These instructions
will call the registered callable before executing the original instruction.

Note that this can interfere with specialization, which will result in
performance degradation in addition to the overhead of calling the
registered callable.

Implementing tools
==================

It is the philosophy of this PEP that third-party tools should be able to
achieve high-performance, not that it should be easy for them to do so.
This PEP provides the necessary API for tools, but does nothing to help 
them determine when and where to insert instrumentation or monitors.

Debuggers
---------

Inserting breakpoints
'''''''''''''''''''''

Breakpoints should be implemented as monitors.
To insert a breakpoint at a given line, the matching instruction offsets
should be found from ``codeobject.co_lines()``.

> Then a monitor should be added for each of those offsets.

Reviewer 2: I'm not super familiar with co_lines(). When does it return multiple offsets? Only when the code of that line is duplicated (e.g. by loop unrolling or similar optimizations)?

To avoid excessive overhead, a single call should be made to
``instrumentation.insert_monitors`` passing all the offsets at once.

> Breakpoints can suspended with ``instrumentation.monitor_off``.

Reviewer 2: Does that re-specialize the instruction at that offset?

Regardless of the answer it would seem this is *deleting* a breakpoint, not just *suspending* it. (If there was a difference, I'd thing that suspending makes it easy to re-enable, but there doesn't seem to be a semantic difference in this case, so I prefer the simpler "deleting".)

Author: > Does that re-specialize the instruction at that offset?

Maybe, but probably not.

Turning a monitor off and removing it have very different performance profiles.

Turning a monitor off is cheap, but will interfere with optimization.
Removing a monitor is very expensive, but will ultimately allow full optimization.

It is up to the implementer of the debugger which to use, but if the user of the debugger turns a breakpoint off, rather than removing it, it implies they may want to turn it on again soon.

Debuggers can break on exceptions being raised by registering a callable 

> for ``RAISE``:

Reviewer 2: So the `RAISE` event applies to *any* reason an exception is raised, not just a `raise` statement, right? This could bear clarification.

``instrumentation.register(RAISE, break_on_raise_handler)``

Stepping
''''''''

Debuggers usually offer the ability to step execution by a
single instruction or line.

This can be implemented by inserting a new monitor at the required
offset(s) of the code to be stepped to,
and by removing or disabling the current monitor.

It is the job of the debugger to compute the relevant offset(s).

Coverage Tools
--------------

Coverage tools need to track which parts of the control graph have been
executed. To do this, they need to track most events and map those events
onto the control flow graph of the code object.
``BRANCH``, ``JUMPBACK``, ``START`` and ``RESUME`` events will inform which
basic blocks have started to execute.

> The ``RAISE`` event with mark any blocks that did not complete.

Reviewer 4: ```suggestion
The ``RAISE`` event will mark any blocks that did not complete.
```

> The ``RAISE`` event with mark any blocks that did not complete.

Reviewer 2: "with"?

This can be then be converted back into a line based report after execution
has completed.

Profilers
---------

Simple profilers need to gather information about calls.
To do this profilers should register for the following events:

> * ENTER
> * EXIT
> * UNWIND
> * C_CALL
> * C_RETURN
> * RAISE

Reviewer 4: ```suggestion
* ``ENTER``
* ``EXIT``
* ``UNWIND``
* ``C_CALL``
* ``C_RETURN``
* ``RAISE``
```

> * RAISE

Reviewer 2: Why is `RAISE` needed here?

Author: To match `C_CALL` the callee raises.

Line based profilers
''''''''''''''''''''

Line based profilers will also need to handle ``BRANCH`` and ``JUMPBACK``
events.
Beware that handling these extra events will have a large performance impact.

.. note::

  Instrumenting profilers have a significant overhead and will distort the
  results of profiling. Unless you need exact call counts,
  consider using a statistical profiler.

Open Issues
===========

[Any points that are still being decided/discussed.]

References
==========

> [A collection of URLs used as references through the PEP.]

Reviewer 2: You can remove this if you're not planning to add any references.
